server/testdriver/core/src/main/java/org/infinispan/server/test/core/rollingupgrade/UpgradeHandler.java

```diff
@@ -0,0 +1,186 @@
+package org.infinispan.server.test.core.rollingupgrade;
+
+import java.net.SocketAddress;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.client.hotrod.RemoteCacheManager;
+import org.infinispan.server.Server;
+import org.infinispan.server.test.api.TestUser;
+import org.infinispan.server.test.core.ContainerInfinispanServerDriver;
+import org.infinispan.server.test.core.InfinispanServerTestConfiguration;
+import org.infinispan.server.test.core.ServerConfigBuilder;
+import org.infinispan.server.test.core.ServerRunMode;
+import org.infinispan.server.test.core.TestSystemPropertyNames;
+
+public class UpgradeHandler {
+   private final UpgradeConfiguration configuration;
+   private final Consumer<String> logConsumer;
+
+   private String toImageCreated;
+   private String fromImageCreated;
+
+   private ContainerInfinispanServerDriver fromDriver;
+   private ContainerInfinispanServerDriver toDriver;
+
+   private UpgradeHandler(UpgradeConfiguration configuration) {
+      this.configuration = configuration;
+      this.logConsumer = configuration.logConsumer();
+   }
+
+   public String getToImageCreated() {
+      return toImageCreated;
+   }
+
+   public String getFromImageCreated() {
+      return fromImageCreated;
+   }
+
+   public ContainerInfinispanServerDriver getFromDriver() {
+      return fromDriver;
+   }
+
+   public ContainerInfinispanServerDriver getToDriver() {
+      return toDriver;
+   }
+
+   public UpgradeConfiguration getConfiguration() {
+      return configuration;
+   }
+
+   public static void performUpgrade(UpgradeConfiguration configuration) throws InterruptedException {
+      String site1Name = "site1";
+      UpgradeHandler handler = new UpgradeHandler(configuration);
+
+      int nodeCount = configuration.nodeCount();
+      String versionFrom = configuration.fromVersion();
+      String versionTo = configuration.toVersion();
+      try {
+         handler.logConsumer.accept("Starting " + nodeCount + " node to version " + versionFrom);
+         handler.fromDriver = handler.startNode(false, configuration.nodeCount(), configuration.nodeCount(),
+               site1Name, configuration.jgroupsProtocol());
+
+         try (RemoteCacheManager manager = handler.createRemoteCacheManager()) {
+            RemoteCache<String, String> cache = configuration.initialHandler().apply(manager);
+
+            for (int i = 0; i < nodeCount; ++i) {
+
+               // TODO: disable state transfer
```

Comment:
Why? In the case of a StatefulSet rolling update and `num_owners > 1`, we need state to be transferred to the node started with the new version.

Reply:
Sorry, this was a comment when I was adding steps. The idea was to have a step where we disable state transfer then add the node then perform a check, then enable state transfer. So that way we can verify we are running commands in between each version and also try to have a better chance of seeing operations during the state transfer.